File: scripts/sortByRef.pl
#!/usr/bin/perl -w

=for comment
Copyright (c) 2012 The Broad Institute

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
THE USE OR OTHER DEALINGS IN THE SOFTWARE.

https://github.com/broadgsa/gatk
=end comment
=cut

use strict;
use Getopt::Long;

sub usage {

    print "\nUsage:\n";
    print "sortByRef.pl [--k POS] [--tmp dir] INPUT REF_DICT\n\n";

    print " Sorts lines of the input file INFILE according\n";
    print " to the reference contig order specified by the\n";
    print " reference dictionary REF_DICT (.fai file).\n";
    print " The sort is stable. If -k option is not specified,\n";
    print " it is assumed that the contig name is the first\n";
    print " field in each line.\n\n";
    print "  INPUT       input file to sort. If '-' is specified, \n";
    print "              then reads from STDIN.\n";
    print "  REF_DICT    .fai file, or ANY file that has contigs, in the\n";
    print "              desired soting order, as its first column.\n";
    print "  --k POS :   contig name is in the field POS (1-based)\n";
    print "              of input lines.\n\n";
    print "  --tmp DIR : temp directory [default=/tmp]\n\n";

    exit(1);
}

my $pos = 1;
my $tmp = "/tmp";
GetOptions( "k:i" => \$pos,
	    "tmp=s" => \$tmp);

$pos--;

usage() if ( scalar(@ARGV) == 0 );

if ( scalar(@ARGV) != 2 ) {
    print "Wrong number of arguments\n";
    usage();
}

my $input_file = $ARGV[0];
my $dict_file = $ARGV[1];


open(DICT, "< $dict_file") or die("Can not open $dict_file: $!");

my %ref_order;

my $n = 0;
while ( <DICT> ) {
    chomp;
    my ($contig, $rest) = split '\s';
    die("Dictionary file is probably corrupt: multiple instances of contig $contig") if ( defined $ref_order{$contig} );

    $ref_order{$contig} = $n;
    $n++;
}

close DICT;
#we have loaded contig ordering now

my $INPUT;
if ($input_file eq "-" ) {
    $INPUT = "STDIN";
} else {
    open($INPUT, "< $input_file") or die("Can not open $input_file: $!");
}

my %temp_outputs;

while ( <$INPUT> ) {
    
    my @fields = split '\s';
    die("Specified field position exceeds the number of fields:\n$_") 
        if ( $pos >= scalar(@fields) );

    my $contig = $fields[$pos];
    if ( $contig =~ m/:/ ) {
        my @loc = split(/:/, $contig);
        # print $contig . " " . $loc[0] . "\n";
        $contig = $loc[0]
    }
    chomp $contig if ( $pos == scalar(@fields) - 1 ); # if last field in line

    my $order;
    if ( defined $ref_order{$contig} ) { $order = $ref_order{$contig}; }
    else {
        $ref_order{$contig} = $n;
        $order = $n; # input line has contig that was not in the dict; 
        $n++; # this contig will go at the end of the output, 
              # after all known contigs
    }

    my $fhandle;
    if ( defined $temp_outputs{$order} ) { $fhandle = $temp_outputs{$order} }
    else {
        #print "opening $order $$ $_\n";
        open( $fhandle, " > $tmp/sortByRef.$$.$order.tmp" ) or
            die ( "Can not open temporary file $order: $!");
        $temp_outputs{$order} = $fhandle;
    }

    # we got the handle to the temp file that keeps all 
    # lines with contig $contig

    print $fhandle $_; # send current line to its corresponding temp file
}

close $INPUT;

foreach my $f ( values %temp_outputs ) { close $f; }

# now collect back into single output stream:

for ( my $i = 0 ; $i < $n ; $i++ ) {
    # if we did not have any lines on contig $i, then there's 
    # no temp file and nothing to do
    next if ( ! defined $temp_outputs{$i} ) ; 

    my $f; 
    open ( $f, "< $tmp/sortByRef.$$.$i.tmp" );
    while ( <$f> ) { print ; }
    close $f;

    unlink "$tmp/sortByRef.$$.$i.tmp";    
}

File: scripts/sortByRef.py
#!/usr/bin/env python

#Copyright (c) 2013, Regents of the University of California
#All rights reserved.
#
#Redistribution and use in source and binary forms, with or without
#modification, are permitted provided that the following conditions are met:
#
#1. Redistributions of source code must retain the above copyright notice,
#this list of conditions and the following disclaimer.
#
#2. Redistributions in binary form must reproduce the above copyright notice,
#this list of conditions and the following disclaimer in the documentation
#and/or other materials provided with the distribution.
#
#THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
#DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
#FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
#DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
#SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
#CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
#OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
#OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import argparse

def main():
    """ 
    Sorts lines of the input file INFILE according
    to the reference contig order specified by the
    reference dictionary REF_DICT (.fai file).
    The sort is stable. If --pos option is not specified,
    it is assumed that the contig name is the first
    field in each line.
    """
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--pos', action='store', type=int, default=0)
    parser.add_argument('--tmp_dir', action='store', type=str, default='/tmp')
    parser.add_argument('input_file', action='store')
    parser.add_argument('output_file', action='store')
    parser.add_argument('ref_dict', action='store')
    args = parser.parse_args()

    # using dict and index is hacky 
    ref_dict = {} 
    index = 0 
    with open(args.ref_dict) as ref_file:
        for line in ref_file:
            contig = line.split('\t')[0] 
            ref_dict[contig] = index
            index = index + 1
    if args.input_file == '-':
        temp_outputs = {}
        in_file = sys.stdin 
        for line in in_file:
            fields = line.split('\t')
            contig = fields[args.pos].split(':')[0]
            if contig in ref_dict:
                order = ref_dict[contig]
            else:
                ref_dict[contig] = index
                order = index
                index = index + 1
            if order in temp_outputs: 
                tmp_file_handle = temp_outputs[order]
            else:
                tmp_name = args.tmp_dir + '/sortByRef_{}.tmp'.format(order)
                tmp_file_handle = open(tmp_name, 'w') 
                temp_outputs[order] = tmp_file_handle
            tmp_file_handle.write(line)
    else:
        temp_outputs = {}
        with open(args.input_file) as in_file:
            for line in in_file:
                fields = line.split('\t')
                contig = fields[args.pos].split(':')[0]
                if contig in ref_dict:
                    order = ref_dict[contig]
                else:
                    ref_dict[contig] = index
                    order = index
                    index = index + 1
                if order in temp_outputs: 
                    tmp_file_handle = temp_outputs[order]
                else:
                    tmp_name = args.tmp_dir + '/sortByRef_{}.tmp'.format(order)
                    tmp_file_handle = open(tmp_name, 'w') 
                    temp_outputs[order] = tmp_file_handle
                tmp_file_handle.write(line)

    for file_handle in temp_outputs.values():
        file_handle.close()

    if args.output_file == '-':
        out_file = sys.stdout 
        for order_num in range(index):
            if order_num in temp_outputs: 
                tmp_name = args.tmp_dir + '/sortByRef_{}.tmp'.format(order_num)
                with open(tmp_name) as tmp_file:
                    out_file.write(tmp_file.read())
                os.remove(tmp_name)                    
            else:
                continue
    else: 
        with open(args.output_file, 'w') as out_file:
            for order_num in range(index):
                if order_num in temp_outputs: 
                    tmp_name = args.tmp_dir + '/sortByRef_{}.tmp'.format(order_num)
                    with open(tmp_name) as tmp_file:
                        out_file.write(tmp_file.read())
                    os.remove(tmp_name)                    
                else:
                    continue


if __name__ == '__main__':
    main()
